modules/javafx.graphics/src/main/java/com/sun/javafx/css/ImmutablePseudoClassSetsCache.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package com.sun.javafx.css;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+import javafx.css.PseudoClass;
+
+/**
+ * A cache for immutable sets of {@link PseudoClass}es.
+ */
+public class ImmutablePseudoClassSetsCache {
+    private static final Map<Set<PseudoClass>, Set<PseudoClass>> CACHE = new HashMap<>();
+
+    /**
+     * Returns an immutable set of {@link PseudoClass}es.
+     * <p>
+     * Note: this method may or may not return the same instance for the same set of
+     * {@link PseudoClass}es.
+     *
+     * @param pseudoClasses a set of {@link PseudoClass} to make immutable, cannot be {@code null}
+     * @return an immutable set of {@link PseudoClass}es, never {@code null}
+     */
+    public static Set<PseudoClass> of(Set<PseudoClass> pseudoClasses) {
+        Set<PseudoClass> immutableSet = Set.copyOf(pseudoClasses);
+
+        /*
+         * Note, the computeIfAbsent looks a bit odd, but it does serve a purpose.
+         *
+         * If the given set was already immutable; the copyOf will return the same
+         * set; however, that may or may not be the set that is already in the cache.
```

Comment:
Do we want to depend on this behavior? The specification of `Set.copyOf` says:
```
@implNote
If the given Collection is an unmodifiable Set, calling copyOf will generally not create a copy.
```

While the actual implementation _never_ creates a copy if the given collection is immutable, the specification does not guarantee that in all cases, but only in the general case. This behavior could theoretically change without changing the specification of `Set.copyOf`.

Reply:
> Do we want to depend on this behavior? The specification of `Set.copyOf` says:
> 
> ```
> @implNote
> If the given Collection is an unmodifiable Set, calling copyOf will generally not create a copy.
> ```
> 
> While the actual implementation _never_ creates a copy if the given collection is immutable, the specification does not guarantee that in all cases, but only in the general case. This behavior could theoretically change without changing the specification of `Set.copyOf`.

If it ever changes implementation, I suppose we could do this logic ourselves -- however, you did give me an idea... before making the copy, I could check if it is in the cache already.  In the case that the set you pass is mutable, but I have an immutable copy already then no copy would need to be made.  The only time a copy is made then is if it wasn't in the cache yet.

Before arriving at this solution I had a class `ImmutablePseudoClassState` which extended `AbstractSet`.  But after stripping everything out that I didn't need, the only thing that was left was a single private field... which was also a `Set`.  I then concluded that there is no need for this wrapper at all, as everything in the CSS code can deal with `Set<PseudoClass>` directly.

Reply:
I've looked into this part a bit further, and I noticed that `BitSet` is also violating the equals/hashCode contract that are supposed to apply to all `Set` implementations.  As I'm now using a `Set` as key for the cache, this contract needs to be adhered to.  I'll see if I can test if this is indeed causing problems as I suspect, and fix it if so.